api/service/log_service.go

```diff
@@ -15,53 +15,290 @@
 package service
 
 import (
+	"bufio"
 	"fmt"
+	"hash/fnv"
 	"io"
+	"sort"
+	"strings"
+	"sync"
 	"time"
 
-	"github.com/gojek/merlin/models"
-	"github.com/google/uuid"
+	"github.com/fatih/color"
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	"github.com/gojek/merlin/cluster"
+	"github.com/gojek/merlin/log"
+	"github.com/gojek/merlin/models"
+)
+
+const (
+	ImageBuilderLabelKey       = "job-name"
+	KnativeServiceLabelKey     = "serving.knative.dev/service"
+	BatchPredictionJobLabelKey = "prediction-job-id"
+
+	prefixAddPod          = "pod"
+	prefixAddContainer    = "container"
+	prefixAddPodContainer = "pod_and_container"
+)
+
+var (
+	skippedContainers = []string{"queue-proxy"}
 )
 
+// LogLine represents a single log line from a container running in Kubernetes.
+type LogLine struct {
+	// Log timestamp in RFC3339 format
+	Timestamp time.Time `json:"timestamp"`
+	// Kubernetes namespace where the pod running the container is created
+	Namespace string `json:"namespace"`
+	// Pod name running the container that produces this log
+	PodName string `json:"pod_name"`
+	// Container name that produces this log
+	ContainerName string `json:"container_name"`
+	// Log in text format
+	TextPayload string `json:"text_payload,omitempty"`
+
+	PrefixColor *color.Color `json:"-"`
+}
+
+func (l LogLine) GenerateText(options LogQuery) string {
+	text := []string{}
+
+	if options.Prefix != "" {
+		p := l.PrefixColor.SprintFunc()
+		if options.Prefix == prefixAddPodContainer {
+			text = append(text, p(l.PodName)+" "+p(l.ContainerName))
+		} else if options.Prefix == prefixAddPod {
+			text = append(text, p(l.PodName))
+		} else if options.Prefix == prefixAddContainer {
+			text = append(text, p(l.ContainerName))
+		}
+	}
+
+	text = append(text, l.Timestamp.Format(time.RFC3339), l.TextPayload)
+	return strings.Join(text, " ") + "\n"
```

Comment:
I think you need to benchmark this function and for comparison maybe using bytes.Buffer to concatenate sting to reduce allocation. For some language time formatting is expensive, I'm not sure with golang better to check also

Reply:
The difference between using strings.Join and bytes.Buffer is not negligible and the `generateText()`'s duration itself is insignificant compare to other functions (fetching the container logs). I'm going to keep it as is.